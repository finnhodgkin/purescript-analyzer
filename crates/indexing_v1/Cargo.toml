[package]
name = "indexing_v1"
version = "0.1.0"
edition = "2021"

[dependencies]
hashbrown = "0.15.2"
id = { version = "0.1.0", path = "../id" }
indexmap = "2.7.0"
insta = "1.42.0"
itertools = "0.14.0"
la-arena = "0.3.1"
petgraph = "0.7.0"
rowan = "0.16.1"
rustc-hash = "2.1.0"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }

[dev-dependencies]
lexing = { path = "../lexing" }
parsing = { path = "../parsing" }
