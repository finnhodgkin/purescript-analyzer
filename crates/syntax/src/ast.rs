#[macro_use]
mod macros;

mod binder;
mod binding;
mod common;
mod declaration;
mod module;
mod name;

pub use binder::*;
pub use binding::*;
pub use common::*;
pub use declaration::*;
pub use module::*;
pub use name::*;
