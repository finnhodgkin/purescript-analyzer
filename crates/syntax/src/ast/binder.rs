use rowan::ast::AstNode;

use crate::SyntaxToken;

use super::{Name, OneOrMore, QualifiedName};

_create_ast_v!(
    Binder,
    ConstructorBinder(ConstructorBinder),
    LiteralBinder(LiteralBinder),
    NegativeBinder(NegativeBinder),
    ParenthesizedBinder(ParenthesizedBinder),
    TypedBinder(TypedBinder),
    VariableBinder(VariableBinder),
    WildcardBinder(WildcardBinder)
);

impl ConstructorBinder {
    pub fn qualified_name(&self) -> Option<QualifiedName> {
        QualifiedName::cast(self.node.first_child()?)
    }

    pub fn fields(&self) -> Option<OneOrMore<Binder>> {
        OneOrMore::cast(self.node.children().nth(1)?)
    }
}

impl ParenthesizedBinder {
    pub fn binder(&self) -> Option<Binder> {
        Binder::cast(self.node.first_child()?)
    }
}

impl NegativeBinder {
    pub fn minus(&self) -> Option<SyntaxToken> {
        self.node.first_token()
    }

    pub fn literal(&self) -> Option<LiteralBinder> {
        LiteralBinder::cast(self.node.first_child()?)
    }
}

impl VariableBinder {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }
}
