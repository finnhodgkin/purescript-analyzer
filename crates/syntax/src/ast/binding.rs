use rowan::ast::AstNode;

use super::OneOrMore;

_create_ast_v!(Binding, UnconditionalBinding(UnconditionalBinding), GuardedBinding(GuardedBinding));

_create_ast!(GuardedExpression, WhereExpression);

impl UnconditionalBinding {
    pub fn where_expression(&self) -> Option<WhereExpression> {
        WhereExpression::cast(self.node.first_child()?)
    }
}

impl GuardedBinding {
    pub fn guarded_expressions(&self) -> Option<OneOrMore<GuardedExpression>> {
        OneOrMore::cast(self.node.first_child()?)
    }
}

impl GuardedExpression {
    pub fn where_expression(&self) -> Option<WhereExpression> {
        WhereExpression::cast(self.node.last_child()?)
    }
}
