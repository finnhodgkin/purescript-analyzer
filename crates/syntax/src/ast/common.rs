use rowan::ast::{support, AstChildren, AstNode};

use crate::PureScript;

_create_ast_t!(OneOrMore, Source, Wrapped);

impl<T: AstNode<Language = PureScript>> OneOrMore<T> {
    pub fn children(&self) -> AstChildren<T> {
        support::children(&self.node)
    }
}

impl<T: AstNode<Language = PureScript>> Source<T> {
    pub fn child(&self) -> Option<T> {
        support::child(&self.node)
    }
}

impl<T: AstNode<Language = PureScript>> Wrapped<T> {
    pub fn child(&self) -> Option<T> {
        support::child(&self.node)
    }
}

_create_ast_t!(ArgumentList, LayoutList);

impl<T: AstNode<Language = PureScript>> ArgumentList<T> {
    pub fn children(&self) -> AstChildren<T> {
        support::children(&self.node)
    }
}

impl<T: AstNode<Language = PureScript>> LayoutList<T> {
    pub fn children(&self) -> AstChildren<T> {
        support::children(&self.node)
    }
}
