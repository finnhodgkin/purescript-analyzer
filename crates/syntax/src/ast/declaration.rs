use rowan::ast::AstNode;

use super::{Binder, Binding, Name, Type, ZeroOrMore};

_create_ast_v!(
    Declaration,
    AnnotationDeclaration(AnnotationDeclaration),
    DataDeclaration(DataDeclaration),
    ForeignDataDeclaration(ForeignDataDeclaration),
    ValueDeclaration(ValueDeclaration)
);

_create_ast_v!(
    LetBinding,
    LetBindingName(LetBindingName),
    LetBindingPattern(LetBindingPattern),
    LetBindingSignature(LetBindingSignature)
);

impl AnnotationDeclaration {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }
}

impl DataDeclaration {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }
}

impl ForeignDataDeclaration {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }

    pub fn ty(&self) -> Option<Type> {
        Type::cast(self.node.last_child()?)
    }
}

impl ValueDeclaration {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }

    pub fn binders(&self) -> Option<ZeroOrMore<Binder>> {
        ZeroOrMore::cast(self.node.children().nth(1)?)
    }

    pub fn binding(&self) -> Option<Binding> {
        Binding::cast(self.node.last_child()?)
    }
}

impl LetBindingName {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }

    pub fn binders(&self) -> Option<ZeroOrMore<Binder>> {
        ZeroOrMore::cast(self.node.children().nth(1)?)
    }

    pub fn binding(&self) -> Option<Binding> {
        Binding::cast(self.node.last_child()?)
    }
}

impl LetBindingSignature {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }
}
