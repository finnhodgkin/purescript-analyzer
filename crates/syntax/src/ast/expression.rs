use rowan::ast::AstNode;

use crate::{PureScript, SyntaxToken};

use super::{LetBinding, Name, NameRef, OneOrMore, QualifiedName, Type};

_create_ast_v!(
    Expression,
    AdoExpression(AdoExpression),
    ApplicationExpression(ApplicationExpression),
    CaseExpression(CaseExpression),
    ConstructorExpression(ConstructorExpression),
    DoExpression(DoExpression),
    ExpressionInfixChain(ExpressionInfixChain),
    ExpressionOperatorChain(ExpressionOperatorChain),
    IfThenElseExpression(IfThenElseExpression),
    LetInExpression(LetInExpression),
    LiteralExpression(LiteralExpression),
    OperatorNameExpression(OperatorNameExpression),
    ParenthesizedExpression(ParenthesizedExpression),
    RecordAccessExpression(RecordAccessExpression),
    RecordUpdateExpression(RecordUpdateExpression),
    TypedExpression(TypedExpression),
    VariableExpression(VariableExpression)
);

_create_ast_v!(Argument, TermArgument(TermArgument), TypeArgument(TypeArgument));

_create_ast_v!(RecordItem, RecordField(RecordField), RecordPun(RecordPun));

impl ApplicationExpression {
    pub fn head(&self) -> Option<Expression> {
        Expression::cast(self.node.first_child()?)
    }

    pub fn spine(&self) -> Option<OneOrMore<Argument>> {
        OneOrMore::cast(self.node.last_child()?)
    }
}

impl TermArgument {
    pub fn expression(&self) -> Option<Expression> {
        Expression::cast(self.node.first_child()?)
    }
}

impl TypeArgument {
    pub fn at(&self) -> Option<SyntaxToken> {
        self.node.first_token()
    }

    pub fn ty(&self) -> Option<Type> {
        Type::cast(self.node.first_child()?)
    }
}

impl LetInExpression {
    pub fn bindings(&self) -> Option<OneOrMore<LetBinding>> {
        OneOrMore::cast(self.node.first_child()?)
    }

    pub fn body(&self) -> Option<Expression> {
        Expression::cast(self.node.last_child()?)
    }
}

impl RecordField {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }

    pub fn value<T>(&self) -> Option<T>
    where
        T: AstNode<Language = PureScript>,
    {
        T::cast(self.node.last_child()?)
    }
}

impl RecordPun {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }

    pub fn name_ref(&self) -> Option<NameRef> {
        NameRef::cast(self.node.first_child()?)
    }
}

impl VariableExpression {
    pub fn qualified_name(&self) -> Option<QualifiedName> {
        QualifiedName::cast(self.node.first_child()?)
    }
}
