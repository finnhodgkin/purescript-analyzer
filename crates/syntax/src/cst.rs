use rowan::ast::{AstChildren, AstNode};

use crate::{SyntaxKind, SyntaxToken};

#[macro_use]
mod macros;

create_cst_struct!(Module, ModuleHeader, ExportList);

create_cst_enum!(
    ExportItem
        | ExportValue
        | ExportClass
        | ExportType
        | ExportOperator
        | ExportTypeOperator
        | ExportModule
);

create_cst_struct!(ModuleImports, ImportStatement, ImportList, ImportAlias);

create_cst_enum!(
    ImportItem // Comment to fix formatting :P
        | ImportValue
        | ImportClass
        | ImportType
        | ImportOperator
        | ImportTypeOperator
);

create_cst_enum!(TypeItems | TypeItemsAll | TypeItemsList);

create_cst_struct!(ModuleStatements);

create_cst_enum!(
    Declaration
        | ValueSignature
        | ValueEquation
        | InfixDeclaration
        | TypeRoleDeclaration
        | InstanceChain
        | InstanceDeclaration
        | InstanceSignatureStatement
        | InstanceEquationStatement
        | TypeSynonymSignature
        | TypeSynonymEquation
        | ClassSignature
        | ClassDeclaration
        | ClassMemberStatement
        | ForeignImportDataDeclaration
        | ForeignImportValueDeclaration
        | NewtypeSignature
        | NewtypeEquation
        | DataSignature
        | DataEquation
        | DeriveDeclaration
);

associated_declarations!(
    ClassDeclaration where ClassMemberStatement
);

associated_declarations!(
    InstanceDeclaration where InstanceSignatureStatement | InstanceEquationStatement
);

create_cst_struct!(
    TypeRole,
    ClassConstraints,
    ClassHead,
    ClassFunctionalDependencies,
    ClassStatements
);

create_cst_enum!(
    FunctionalDependency | FunctionalDependencyDetermined | FunctionalDependencyDetermines
);

create_cst_struct!(InstanceName, InstanceConstraints, InstanceHead, InstanceStatements);

create_cst_struct!(DataConstructor);

create_cst_enum!(
    Type // Comment to fix formatting :P
        | TypeApplicationChain
        | TypeArrow
        | TypeConstrained
        | TypeConstructor
        | TypeForall
        | TypeHole
        | TypeInteger
        | TypeKinded
        | TypeOperator
        | TypeOperatorChain
        | TypeString
        | TypeVariable
        | TypeVariableBinding
        | TypeWildcard
        | TypeRecord
        | TypeRow
        | TypeParenthesized
);

create_cst_enum!(
    Binder
        | BinderTyped
        | BinderOperatorChain
        | BinderInteger
        | BinderNumber
        | BinderConstructor
        | BinderVariable
        | BinderNamed
        | BinderWildcard
        | BinderString
        | BinderChar
        | BinderTrue
        | BinderFalse
        | BinderArray
        | BinderRecord
        | BinderParenthesized
);

create_cst_struct!(
    FunctionBinders,
    Unconditional,
    WhereExpression,
    Conditionals,
    PatternGuarded,
    PatternGuards
);

create_cst_enum!(PatternGuard | PatternGuardBinder | PatternGuardExpression);

create_cst_enum!(
    Expression
        | ExpressionTyped
        | ExpressionOperatorChain
        | ExpressionInfixChain
        | ExpressionTick
        | ExpressionNegate
        | ExpressionApplicationChain
        | ExpressionTypeArgument
        | ExpressionTermArgument
        | ExpressionIfThenElse
        | ExpressionLetIn
        | ExpressionLambda
        | ExpressionCaseOf
        | ExpressionDo
        | ExpressionAdo
        | ExpressionConstructor
        | ExpressionVariable
        | ExpressionOperatorName
        | ExpressionSection
        | ExpressionHole
        | ExpressionString
        | ExpressionChar
        | ExpressionTrue
        | ExpressionFalse
        | ExpressionInteger
        | ExpressionNumber
        | ExpressionArray
        | ExpressionRecord
        | ExpressionParenthesized
        | ExpressionRecordAccess
        | ExpressionRecordUpdate
);

create_cst_struct!(CaseTrunk, CaseBranches, CaseBranchBinders, CaseBranch);

create_cst_enum!(DoStatement | DoStatementBind | DoStatementLet | DoStatementDiscard);

create_cst_enum!(LetBinding | LetBindingPattern | LetBindingSignature | LetBindingEquation);

create_cst_enum!(RecordItem | RecordField | RecordPun);

create_cst_struct!(RecordUpdates);

create_cst_enum!(RecordUpdate | RecordUpdateLeaf | RecordUpdateBranch);

impl Module {
    pub fn statements(&self) -> Option<ModuleStatements> {
        self.syntax().children().find_map(ModuleStatements::cast)
    }
}

impl ModuleStatements {
    pub fn children(&self) -> impl Iterator<Item = Declaration> {
        self.syntax().children().filter_map(Declaration::cast)
    }
}

impl InstanceChain {
    pub fn instance_declarations(&self) -> AstChildren<InstanceDeclaration> {
        rowan::ast::support::children(self.syntax())
    }
}

impl InstanceDeclaration {
    pub fn instance_head(&self) -> Option<InstanceHead> {
        rowan::ast::support::child(self.syntax())
    }

    pub fn instance_name(&self) -> Option<InstanceName> {
        rowan::ast::support::child(self.syntax())
    }

    pub fn instance_statements(&self) -> Option<InstanceStatements> {
        rowan::ast::support::child(self.syntax())
    }
}

impl InstanceName {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}

impl InstanceHead {
    pub fn type_name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::UPPER)
    }
}

impl InstanceStatements {
    pub fn children(&self) -> AstChildren<InstanceDeclarationStatement> {
        rowan::ast::support::children(self.syntax())
    }
}

impl ValueSignature {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}

impl ValueEquation {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}
