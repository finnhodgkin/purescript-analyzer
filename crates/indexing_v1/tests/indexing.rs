use std::fmt::Write;

use indexing_v1::{IndexingErrors, IndexingResult};
use rowan::ast::AstNode;
use syntax::cst;

fn index_source<'s>(
    source: impl AsRef<[&'s str]>,
) -> (cst::Module, IndexingResult, IndexingErrors) {
    let source = source.as_ref().join("\n");

    let lexed = lexing::lex(&source);
    let tokens = lexing::layout(&lexed);

    let (module, _) = parsing::parse(&lexed, &tokens);
    let module = cst::Module::cast(module).unwrap();

    let (index, errors) = indexing_v1::index(&module);
    (module, index, errors)
}

fn index<'s>(lines: impl AsRef<[&'s str]>) -> (cst::Module, IndexingResult, IndexingErrors) {
    let source = format!("module Main where\n{}", lines.as_ref().join("\n"));
    index_source([source.as_str()])
}

#[test]
fn well_formed_module() {
    let (_, index, _) = index([
        "import Prelude as Prelude",
        "import Data.List as Data.List",
        "import Halogen as H",
        "import Halogen.HTML as H",
        "id :: forall a. a -> a",
        "id x = x",
        "data Maybe :: Type -> Type",
        "data Maybe a = Just a | Nothing",
        "type role Maybe representational",
        "class Eq :: Type -> Constraint",
        "class Eq a where",
        "  eq :: a -> a -> Bool",
        "instance eqInt :: Eq Int where",
        "  eq :: Int -> Int -> Bool",
        "  eq = primEqInt",
        "newtype Identity :: Type -> Type",
        "newtype Identity a = Identity a",
        "type role Identity representational",
        "type IdentityInt :: Type",
        "type IdentityInt = Identity Int",
        "foreign import primEqInt :: Int -> Int -> Bool",
        "foreign import data Functor :: Type -> Type",
        "type role Functor phantom",
        "derive newtype instance eqIdentity :: Eq (Identity a)",
        "infix 5 eq as ==",
        "infix 5 type Eq as ==",
    ]);

    let prelude = index.nominal.lookup_qualified("Prelude");
    let data_list = index.nominal.lookup_qualified("Data.List");
    let halogen = index.nominal.lookup_qualified("H");
    let id = index.nominal.lookup_expr_item("id");
    let maybe = index.nominal.lookup_type_item("Maybe");
    let just = index.nominal.lookup_expr_item("Just");
    let nothing = index.nominal.lookup_expr_item("Nothing");
    let eq_class = index.nominal.lookup_type_item("Eq");
    let eq = index.nominal.lookup_expr_item("eq");
    let eq_int = index.nominal.lookup_expr_item("eqInt");
    let identity = index.nominal.lookup_type_item("Identity");
    let identity_int = index.nominal.lookup_type_item("IdentityInt");
    let prim_eq_int = index.nominal.lookup_expr_item("primEqInt");
    let functor = index.nominal.lookup_type_item("Functor");
    let eq_identity = index.nominal.lookup_expr_item("eqIdentity");
    let expr_equal = index.nominal.lookup_expr_item("==");
    let type_equal = index.nominal.lookup_type_item("==");

    let mut snapshot = String::new();
    writeln!(snapshot, "prelude: {:?}", prelude).unwrap();
    writeln!(snapshot, "data_list: {:?}", data_list).unwrap();
    writeln!(snapshot, "halogen: {:?}", halogen).unwrap();
    writeln!(snapshot, "id: {:?}", id).unwrap();
    writeln!(snapshot, "maybe: {:?}", maybe).unwrap();
    writeln!(snapshot, "just: {:?}", just).unwrap();
    writeln!(snapshot, "nothing: {:?}", nothing).unwrap();
    writeln!(snapshot, "eq_class: {:?}", eq_class).unwrap();
    writeln!(snapshot, "eq: {:?}", eq).unwrap();
    writeln!(snapshot, "eq_int: {:?}", eq_int).unwrap();
    writeln!(snapshot, "identity: {:?}", identity).unwrap();
    writeln!(snapshot, "identity_int: {:?}", identity_int).unwrap();
    writeln!(snapshot, "prim_eq_int: {:?}", prim_eq_int).unwrap();
    writeln!(snapshot, "functor: {:?}", functor).unwrap();
    writeln!(snapshot, "eq_identity: {:?}", eq_identity).unwrap();
    writeln!(snapshot, "expr_equal: {:?}", expr_equal).unwrap();
    writeln!(snapshot, "type_equal: {:?}", type_equal).unwrap();
    insta::assert_snapshot!(snapshot);
}

#[test]
fn duplicate_expr_item() {
    let (_, _, errors) = index([
        "eq :: forall a. a -> a -> Bool",
        "class Eq a where",
        "  eq :: forall a. a -> a -> Bool",
        "instance eq :: Eq a",
        "foreign import eq :: forall a. a -> a -> Bool",
        "instance ord :: Ord a",
        "ord :: forall a. a -> a -> Comparison",
        "ord = ordImpl",
        "data Data = Id Int",
        "newtype Newtype = Id Int",
        "infix 5 eq as ==",
        "infix 5 eq as ==",
        "derive instance eq :: Eq a",
    ]);
    insta::assert_debug_snapshot!(errors);
}

#[test]
fn duplicate_type_item() {
    let (_, _, errors) = index([
        "data Data = Constructor0",
        "newtype Data = Constructor1",
        "class Data",
        "foreign import data Data :: Type",
        "type Data = Type",
        "newtype Newtype = Constructor2",
        "data Newtype = Constructor3",
        "data Newtype :: Type",
        "data Id = Constructor4",
        "data Id = Constructor5",
        "infix 5 type Eq as ==",
        "infix 5 type Eq as ==",
    ]);
    insta::assert_debug_snapshot!(errors);
}

#[test]
fn value_duplicate_signature() {
    let (_, _, errors) = index([
        "isJust :: forall a. Maybe a -> Bool",
        "isJust (Just _) = true",
        "isJust Nothing = false",
        "isJust :: forall a. Maybe a -> Bool",
    ]);
    insta::assert_debug_snapshot!(errors);
}

#[test]
fn value_non_consecutive() {
    let (_, _, errors) = index([
        "isJust :: forall a. Maybe a -> Bool",
        "isJust (Just _) = true",
        "life = 42",
        "isJust Nothing = false",
    ]);
    insta::assert_debug_snapshot!(errors);
}

#[test]
fn value_late_signature() {
    let (_, _, errors) = index([
        "isJust (Just _) = true",
        "isJust Nothing = false",
        "isJust :: forall a. Maybe a -> Bool",
    ]);
    insta::assert_debug_snapshot!(errors);
}

#[test]
fn type_duplicate_signature() {
    let (_, _, errors) = index([
        "data Maybe :: Type -> Type",
        "data Maybe a = Just a | Nothing",
        "data Maybe :: Type -> Type",
    ]);
    insta::assert_debug_snapshot!(errors);
}

#[test]
fn type_non_consecutive() {
    let (_, _, errors) = index([
        "data Maybe :: Type -> Type",
        "life :: Int",
        "life = 42",
        "data Maybe a = Just a | Nothing",
    ]);
    insta::assert_debug_snapshot!(errors);
}

#[test]
fn type_late_signature() {
    let (_, _, errors) = index(["data Maybe a = Just a | Nothing", "data Maybe :: Type -> Type"]);
    insta::assert_debug_snapshot!(errors);
}

#[test]
fn type_role_errors() {
    let (_, _, errors) = index([
        "class Eq a",
        "type role Eq representational",
        "type role Ord representational",
        "newtype Id :: Type -> Type",
        "type role Id nominal",
        "type role Id phantom",
    ]);
    insta::assert_debug_snapshot!(errors);
}

#[test]
fn well_formed_export() {
    let (_, index, _) = index_source([
        "module Main (life, class Eq, Synonym, Data(..), List(Cons, Nil), (+), type (+), module Export) where",
        "import Lib as Export",
        "life = 42", 
        "class Eq", 
        "type Synonym = Int", 
        "data Data = A | B", 
        "data List = Cons | Nil",
        "infix 5 + as add",
        "infix 5 type + as Add",
    ]);

    let life = index.nominal.lookup_expr_item("life");
    let eq_class = index.nominal.lookup_type_item("Eq");
    let synonym = index.nominal.lookup_type_item("Synonym");
    let data = index.nominal.lookup_type_item("Data");
    let data_a = index.nominal.lookup_expr_item("A");
    let data_b = index.nominal.lookup_expr_item("B");
    let list = index.nominal.lookup_type_item("List");
    let list_cons = index.nominal.lookup_expr_item("Cons");
    let list_nil = index.nominal.lookup_expr_item("Nil");
    let plus = index.nominal.lookup_expr_item("+");
    let plus_type = index.nominal.lookup_type_item("+");
    let export = index.nominal.lookup_qualified("Export");

    let mut snapshot = String::default();
    writeln!(snapshot, "life: {:?}", life).unwrap();
    writeln!(snapshot, "eq_class: {:?}", eq_class).unwrap();
    writeln!(snapshot, "synonym: {:?}", synonym).unwrap();
    writeln!(snapshot, "data: {:?}", data).unwrap();
    writeln!(snapshot, "data_a: {:?}", data_a).unwrap();
    writeln!(snapshot, "data_b: {:?}", data_b).unwrap();
    writeln!(snapshot, "list: {:?}", list).unwrap();
    writeln!(snapshot, "list_cons: {:?}", list_cons).unwrap();
    writeln!(snapshot, "list_nil: {:?}", list_nil).unwrap();
    writeln!(snapshot, "plus: {:?}", plus).unwrap();
    writeln!(snapshot, "plus_type: {:?}", plus_type).unwrap();
    writeln!(snapshot, "export: {:?}", export).unwrap();
    insta::assert_snapshot!(snapshot);
}

#[test]
fn duplicate_export_item() {
    let (_, _, errors) = index_source([
        "module Main (life, life, class Eq, class Eq, Synonym, Synonym, Data(..), Data(..), (+), (+), type (+), type (+), module Export, module Export) where",
        "import Lib as Export",
        "life = 42", 
        "class Eq", 
        "type Synonym = Int", 
        "data Data = A | B", 
        "infix 5 + as add",
        "infix 5 type + as Add",
    ]);
    insta::assert_debug_snapshot!(errors);
}

#[test]
fn export_type_item_error() {
    let (_, _, errors) = index_source([
        "module Main (Id(..), Index(Index)) where",
        "type Id = Int",
        "type Index = Int",
    ]);
    insta::assert_debug_snapshot!(errors);
}
