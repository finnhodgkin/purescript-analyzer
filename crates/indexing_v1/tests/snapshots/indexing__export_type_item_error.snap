---
source: crates/indexing_v1/tests/indexing.rs
expression: errors
snapshot_kind: text
---
[
    InvalidTypeItemExport {
        export_id: Id<syntax::cst::ExportItem>(0),
        type_id: Id<indexing_v1::indexes::TypeItem>(0),
    },
    InvalidTypeItemExport {
        export_id: Id<syntax::cst::ExportItem>(1),
        type_id: Id<indexing_v1::indexes::TypeItem>(1),
    },
]
